### PR TITLE
Feature/sensors

### PR DESCRIPTION
**Pull request by trmdy**

### CODE REVIEW

File: sensors/models.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from django.db import models
+
+# Create your models here.
```
Comment (chdefrene) [inline sensors/models.py]:
```suggestion
```

File: sensors/models.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+class SensorMeasurement(models.Model):
+    type = models.CharField(
+        blank=False, null=False, max_length=16, choices=MEASUREMENT_TYPE_CHOICES
+    )
+    value = models.FloatField(blank=False, null=False)
+    created_at = models.DateTimeField(blank=False, null=False, default=timezone.now)
+
+    def __str__(self):
+        return f"{self.get_type_display()} measurement"
```
Comment (chdefrene) [inline sensors/models.py:16]: Where is this method implemented?

File: sensors/static/sensors/js/overview.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+getTemperatureData()
+getHumidityData()
+getSoundData()
+getPeopleData()
+
+setInterval(getTemperatureData, 20 * 1000)
+setInterval(getHumidityData, 20 * 1000)
+setInterval(getSoundData, 20 * 1000)
+setInterval(getPeopleData, 20 * 1000)
```
Comment (chdefrene) [inline sensors/static/sensors/js/overview.js]: This means the functions should be called every 20 seconds, right? Why not have this value defined a a constant?

File: sensors/migrations/0002_auto_20190404_1424.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+    dependencies = [
+        ('sensors', '0001_initial'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='sensormeasurement',
+            name='type',
+            field=models.CharField(choices=[('temperature', 'Temperature'), ('humidity', 'Humidity'), ('sound', 'Sound'), ('people', 'People count')], max_length=16),
```
Comment (chdefrene) [inline sensors/migrations/0002_auto_20190404_1424.py:16]: This could be merged with migration 0001

File: sensors/consts.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+MEASUREMENT_TYPE_HUMIDITY = 'humidity'
+MEASUREMENT_TYPE_SOUND = 'sound'
+MEASUREMENT_TYPE_PEOPLE_COUNT = 'people'
+
+MEASUREMENT_TYPE_CHOICES = (
+    (MEASUREMENT_TYPE_TEMPERATURE, 'Temperature'),
+    (MEASUREMENT_TYPE_HUMIDITY, 'Humidity'),
+    (MEASUREMENT_TYPE_SOUND, 'Sound'),
+    (MEASUREMENT_TYPE_PEOPLE_COUNT, 'People count'),
+)
```
Comment (chdefrene) [inline sensors/consts.py:11]: Please have a look at #80 for a neat(er) implementation of choice enums like this

File: api/views.py [hunk 6]
```diff
@@ -127,3 +132,79 @@ def post(self, request, *args, **kwargs):
         data = serializer.data
 
         return Response(data, status=status.HTTP_201_CREATED)
+
+    @staticmethod
+    def _get_account_from_card_id(card_uuid) -> SociBankAccount:
+        soci_bank_account = get_object_or_404(queryset=SociBankAccount.objects.all(), card_uuid=card_uuid)
+
+        return soci_bank_account
```
Comment (chdefrene) [inline api/views.py]: This was removed in a previous PR. Please rebase before pushing your changes 😛

File: api/views.py [hunk 7]
```diff
@@ -127,3 +132,79 @@ def post(self, request, *args, **kwargs):
...
+        responses={
+            "201": serializer_class(many=True),
+            "400": ": Illegal input.",
+            "403": ": You are not authorized to create measurements."
+        },
+    )
+    def post(self, request: Request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data, many=True)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
```
Comment (chdefrene) [inline api/views.py]: I would prefer distinguishing between deserializing and serializing. I have previously made a mixin that can be used to provide a `deserializer_class`
```suggestion
        deserializer.save()
```

File: api/views.py [hunk 8]
```diff
@@ -127,3 +132,79 @@ def post(self, request, *args, **kwargs):
...
+
+        # Try to clear out old entries.
+        try:
+            SensorMeasurementView._remove_old_measurement_instances()
+        except DatabaseError:
+            # Should we do something?
+            pass
```
Comment (chdefrene) [inline api/views.py:206]: How can this method fail? I cannot immediately see why a filter.delete would result in this error.

File: api/views.py [hunk 9]
```diff
@@ -127,3 +132,79 @@ def post(self, request, *args, **kwargs):
...
+            name='type',
+            in_=IN_QUERY,
+            default="temperature",
+            description=f"The type of measurements.",
+            type=TYPE_STRING,
+            enum=[x[0] for x in MEASUREMENT_TYPE_CHOICES],
+            required=False
+        )],
+        responses={
+            "200": serializer_class(many=True),
```
Comment (chdefrene) [inline api/views.py:222]: I believe that when this is the only response, it will be automatically generated by drf-yasg, Cf. line 42

**Review by chdefrene — CHANGES_REQUESTED** [review]

File: sensors/models.py (hunk 2 above)
Comment (trmdy) [reply sensors/models.py:16]: See https://docs.djangoproject.com/en/2.2/ref/models/instances/#django.db.models.Model.get_FOO_display

File: sensors/migrations/0002_auto_20190404_1424.py (hunk 4 above)
Comment (trmdy) [reply sensors/migrations/0002_auto_20190404_1424.py:16]: It _could_, but optimizing migrations at this stage is somewhat premature.

File: sensors/consts.py (hunk 5 above)
Comment (trmdy) [reply sensors/consts.py:11]: I'm not sure I see the point with that. It's semantically confusing having to rename types into "status", just to invoke a shorthand for CharField(max_length=100, choices=<something>). For me, this is less explicit and harder to read (as you have another level of indirection).

File: api/views.py (hunk 8 above)
Comment (trmdy) [reply api/views.py:206]: Anything related to IO can fail unexpectedly. The `DatabaseError` is a generic catch-all for Django database errors. The reason I have this here is that the clean-up of old measurement instances is not essential for the view to complete successfully. Hence some error occurring should not fail the view.

If we had a general async task processor, this code would run there instead, but for now, I find this to be the best solution.

File: api/views.py (hunk 9 above)
Comment (trmdy) [reply api/views.py:222]: I see. I think I'll let it stay here though — unless you consider that a problem. Having things like this explicitly makes it easier for DRF/Swagger-beginners like myself to see what is supposed to be output.

**Review by alexaor — APPROVED** [review]

**Pull request merged**